File: go/0054_spiral_matrix.go
func spiralOrder(matrix [][]int) []int {
	if len(matrix) == 0 || len(matrix[0]) == 0 {
		return []int{}
	}

	var res []int
	top, bottom, left, right := 0, len(matrix)-1, 0, len(matrix[0])-1
	point := []int{0, 0}

	for {
		// move to right
		for i := point[1]; i <= right; i++ {
			res = append(res, matrix[point[0]][i])
		}
		if point[0] >= bottom {
			break
		}
		top, point[0], point[1] = top+1, point[0]+1, right

		// move to bottom
		for i := point[0]; i <= bottom; i++ {
			res = append(res, matrix[i][point[1]])
		}
		if point[1] <= left {
			break
		}
		right, point[0], point[1] = right-1, bottom, point[1]-1

		// move to left
		for i := point[1]; i >= left; i-- {
			res = append(res, matrix[point[0]][i])
		}
		if point[0] <= top {
			break
		}
		bottom, point[0], point[1] = bottom-1, point[0]-1, left

		// move to top
		for i := point[0]; i >= top; i-- {
			res = append(res, matrix[i][point[1]])
		}
		if point[1] >= right {
			break
		}
		left, point[0], point[1] = left+1, top, point[1]+1
	}
	return res
}

File: python/0054_spiral_matrix.py
class Solution:
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """

        mlen = len(matrix)
        if mlen == 0:
            return []
        nlen = len(matrix[0])
        if nlen == 0:
            return []

        left, right, top, bottom = 0, nlen, 0, mlen
        res = []
        while True:
            for n in range(left, right):
                res.append(matrix[top][n])

            top += 1
            if top >= bottom:
                break

            for m in range(top, bottom):
                res.append(matrix[m][right - 1])

            right -= 1
            if right <= left:
                break

            for n in reversed(range(left, right)):
                res.append(matrix[bottom - 1][n])

            bottom -= 1
            if bottom <= top:
                break

            for m in reversed(range(top, bottom)):
                res.append(matrix[m][left])

            left += 1
            if left >= right:
                break
        return res